tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,244 @@
+			it.cur = val
+			return true
+		}
+	}
+
+	return false
+}
+
+func (it *intersectLabelValuesV1) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValuesV1) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValuesV1) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValuesV1) Close() error {
+	return nil
+}
+
+type intersectLabelValues struct {
+	d           *encoding.Decbuf
+	b           ByteSlice
+	dec         *Decoder
+	postings    Postings
+	curPostings bigEndianPostings
+	lastVal     string
+	skip        int
+	cur         string
+	exhausted   bool
+	err         error
+}
+
+func (it *intersectLabelValues) Next() bool {
+	if it.exhausted {
+		return false
+	}
+
+	for !it.exhausted && it.d.Err() == nil {
+		if it.skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than to parse.
+			it.skip = it.d.Len()
+			// Key count
+			it.d.Uvarint()
+			// Label name
+			it.d.UvarintBytes()
+			it.skip -= it.d.Len()
+		} else {
+			it.d.Skip(it.skip)
+		}
+
+		// Label value
+		v := yoloString(it.d.UvarintBytes())
+
+		postingsOff := it.d.Uvarint64()
+		// Read from the postings table
+		d2 := encoding.NewDecbufAt(it.b, int(postingsOff), castagnoliTable)
+		if _, err := it.dec.PostingsInPlace(d2.Get(), &it.curPostings); err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		it.exhausted = v == it.lastVal
+
+		it.postings.Reset()
+		if checkIntersection(&it.curPostings, it.postings) {
+			it.cur = v
+			return true
+		}
+	}
+	if it.d.Err() != nil {
+		it.err = fmt.Errorf("get postings offset entry: %w", it.d.Err())
+	}
+
+	return false
+}
+
+func (it *intersectLabelValues) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValues) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValues) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValues) Close() error {
+	return nil
+}
+
+// LabelValuesFor returns LabelValues for the given label name in the series referred to by postings.
+func (p *MemPostings) LabelValuesFor(postings Postings, name string) storage.LabelValues {
+	p.mtx.RLock()
+
+	e := p.m[name]
+	if len(e) == 0 {
+		p.mtx.RUnlock()
+		return storage.EmptyLabelValues()
+	}
+
+	// With thread safety in mind and due to random key ordering in map, we have to construct the array in memory
+	vals := make([]string, 0, len(e))
+	candidates := make([]Postings, 0, len(e))
+	for val, srs := range e {
+		vals = append(vals, val)
+		candidates = append(candidates, NewListPostings(srs))
+	}
+	p.mtx.RUnlock()
+
+	indexes, err := FindIntersectingPostings(postings, candidates)
+	if err != nil {
+		return storage.ErrLabelValues(err)
+	}
+
+	// Filter the values, keeping only those with intersecting postings
+	slices.Sort(indexes)
+	for i := range vals {
+		if _, ok := slices.BinarySearch(indexes, i); !ok {
+			vals = slices.Delete(vals, i, i+1)
+		}
+	}
```

Comment:
There's a more efficient way of doing this: just go through the indexes and append on `vals` the `vals[i]` where `i` is each one of kept indexes.

Reply:
I realized my comment is hard to understand, this is what I mean: 
```suggestion
	// Filter the values, keeping only those with intersecting postings
	if len(vals) != len(indexes) {
		slices.Sort(indexes)
		for i, index := range indexes {
		    vals[i] = vals[index]
		}
		vals = vals[:len(indexes)]
	}
```